### PR TITLE
#1353 Create one module for every Building Blocks and Tests

### LINKED ISSUE #1353: When loading a V11 project - a separate module is created for each BB
I have a project with exactly one simulation and one BB of each type. I would expect that all BBs will be in one single module when opening in V12. However, this is not the case.

image

I cannot upload the project here but will send it to you via email.

### PR DESCRIPTION
**Pull request by benjaperez1983**
Fixes #1353

# Description

Create one module for every Building Blocks when they are Unique

## Type of change

Please mark relevant options with an `x` in the brackets.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires documentation changes (link at least one [user](https://github.com/Open-Systems-Pharmacology/docs) or [developer](https://github.com/Open-Systems-Pharmacology/developer-docs) documentation issue):
- [ ] Algorithm update - updates algorithm documentation/questions/answers etc.
- [ ] Other (please describe):

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Integration tests
- [x] Unit tests
- [x] Manual tests 
- [ ] No tests required

# Reviewer checklist

Mark everything that needs to be checked before merging the PR.

- [ ] Check if the code is well documented
- [ ] Check if the behavior is what is expected
- [ ] Check if the code is well tested
- [ ] Check if the code is readable and well formatted
- [ ] Additional checks (document below if any)
- [ ] Check if documentation update issue(s) are created if the option `This change requires a documentation update` above is selected

# Screenshots (if appropriate):

# Questions (if appropriate):

### CODE REVIEW

File: src/MoBi.Core/Serialization/Converter/ModuleConfigurationBuilder.cs [hunk 1]
```diff
@@ -1,45 +1,58 @@
 ﻿using System.Linq;
+using DevExpress.Utils.About;
```
Comment (rwmcintosh) [inline src/MoBi.Core/Serialization/Converter/ModuleConfigurationBuilder.cs:2]: check if these new ones are really used or not. I don't think so.

If you use Extensions -> ReSharper -> Edit -> Cleanup Code...

This will be removed for you.

File: src/MoBi.Core/Serialization/ORM/Mappers/ProjectMetaDataToProjectMapper.cs [hunk 2]
```diff
@@ -91,9 +92,22 @@ private void addNonProjectBuildingBlocks(List<IBuildingBlock> moduleBuildingBloc
          var simulationSettingsBlocks = moduleBuildingBlocks.OfType<SimulationSettings>().ToList();
          simulationSettingsBlocks.Each(x => moduleBuildingBlocks.Remove(x));
 
-         addModuleBuildingBlocks(moduleBuildingBlocks);
+
+         var uniqueBuildingBlockTypes = hasUniqueBuildingBlocks(moduleBuildingBlocks);
```
Comment (rwmcintosh) [inline src/MoBi.Core/Serialization/ORM/Mappers/ProjectMetaDataToProjectMapper.cs:96]: you can use the method directly in the condition

`if(hasUniqueBuildingBlocks(moduleBuildingBlocks))`

File: src/MoBi.Core/Serialization/ORM/Mappers/ProjectMetaDataToProjectMapper.cs [hunk 3]
```diff
@@ -91,9 +92,22 @@ private void addNonProjectBuildingBlocks(List<IBuildingBlock> moduleBuildingBloc
...
+         if(uniqueBuildingBlockTypes)
+            addModuleBuildingBlocks(moduleBuildingBlocks);
+         else
+            addBuildingBlocksToModule(moduleBuildingBlocks);
+         
          addSimulationSettingsBuildingBlocks(simulationSettingsBlocks);
       }
+      private void addBuildingBlocksToModule(List<IBuildingBlock> moduleBuildingBlocks)
+      {
+         if (!moduleBuildingBlocks.Any())
```
Comment (rwmcintosh) [inline src/MoBi.Core/Serialization/ORM/Mappers/ProjectMetaDataToProjectMapper.cs:106]: Is that needed? I think the `Each` implies that if there are none, then nothing will happen

File: src/MoBi.Core/Serialization/ORM/Mappers/ProjectMetaDataToProjectMapper.cs [hunk 4]
```diff
@@ -109,8 +123,9 @@ private void addModuleBuildingBlocks(List<IBuildingBlock> moduleBuildingBlocks)
       {
          if (!moduleBuildingBlocks.Any())
             return;
-
-         moduleBuildingBlocks.Each(addBuildingBlockAsModule);
+         var module = _moduleFactory.CreateModuleWithName("Module1");
```
Comment (rwmcintosh) [inline src/MoBi.Core/Serialization/ORM/Mappers/ProjectMetaDataToProjectMapper.cs:126]: Could you create a property in `AppConstants` called `DefaultSingleModuleName`.

File: src/MoBi.Core/Services/ModuleFactory.cs [hunk 5]
```diff
@@ -20,6 +21,12 @@ public ModuleFactory(IObjectBaseFactory objectBaseFactory, IObjectTypeResolver o
          _objectTypeResolver = objectTypeResolver;
       }
 
+      public Module CreateModuleWithName(string moduleName)
+      {
+         var newModule = _objectBaseFactory.Create<Module>().WithName(moduleName);
```
Comment (rwmcintosh) [inline src/MoBi.Core/Services/ModuleFactory.cs:26]: Return directly instead of creating the variable. You can also change to an expression body when the method has only one line.

File: tests/MoBi.Tests/IntegrationTests/ProjectTaskIntegrationTests.cs [hunk 6]
```diff
@@ -39,15 +52,57 @@ public void should_have_set_the_name_of_the_project_to_the_name_of_the_file()
...
+         simulation.Modules.Each(x => matchSimulationAndProjectModule(x, _context.CurrentProject.Modules));
+      }
+
+      private void matchSimulationAndProjectModule(Module module, IReadOnlyList<Module> projectModules)
+      {
+         projectModules.SingleOrDefault(x => x.IsNamed(module.Name)).ShouldNotBeNull();
+      }
+   }
+
+   public class When_Loading_a_Project_with_Duplicated_Building_Block_Types : concern_for_LoadProjectIntegrationTest
```
Comment (rwmcintosh) [inline tests/MoBi.Tests/IntegrationTests/ProjectTaskIntegrationTests.cs:76]: Nice!

**Comment by rwmcintosh** [comment]
Looking good here. 

Some OSPSuite coding convention stuff for you to look at. I know you also told me that you were not done.

File: src/MoBi.Core/Serialization/ORM/Mappers/ProjectMetaDataToProjectMapper.cs [hunk 7]
```diff
@@ -91,9 +93,18 @@ private void addNonProjectBuildingBlocks(List<IBuildingBlock> moduleBuildingBloc
          var simulationSettingsBlocks = moduleBuildingBlocks.OfType<SimulationSettings>().ToList();
          simulationSettingsBlocks.Each(x => moduleBuildingBlocks.Remove(x));
 
-         addModuleBuildingBlocks(moduleBuildingBlocks);
+
+
```
Comment (rwmcintosh) [inline src/MoBi.Core/Serialization/ORM/Mappers/ProjectMetaDataToProjectMapper.cs:97]: remove the extra whitespace.

File: src/MoBi.Core/Serialization/ORM/Mappers/ProjectMetaDataToProjectMapper.cs [hunk 8]
```diff
@@ -91,9 +93,18 @@ private void addNonProjectBuildingBlocks(List<IBuildingBlock> moduleBuildingBloc
          var simulationSettingsBlocks = moduleBuildingBlocks.OfType<SimulationSettings>().ToList();
          simulationSettingsBlocks.Each(x => moduleBuildingBlocks.Remove(x));
 
-         addModuleBuildingBlocks(moduleBuildingBlocks);
+
+         
+         if(hasUniqueBuildingBlocks(moduleBuildingBlocks))
+            addModuleBuildingBlocks(moduleBuildingBlocks);
+         else
+            addBuildingBlocksToModule(moduleBuildingBlocks);
```
Comment (rwmcintosh) [inline src/MoBi.Core/Serialization/ORM/Mappers/ProjectMetaDataToProjectMapper.cs:101]: can you try to rename these two methods `addModuleBuildingBlocks` and `addBuildingBlocksToModule`. When I am reading along with the code, I can't determine what will be different.

maybe you can name one of them `addAllBuildingBlocksToModule` and one could be `addAllBuildingBlocksToSeparateModules`

File: tests/MoBi.Tests/IntegrationTests/ProjectTaskIntegrationTests.cs [hunk 9]
```diff
@@ -39,15 +52,57 @@ public void should_have_set_the_name_of_the_project_to_the_name_of_the_file()
...
 
       [Observation]
-      public void each_module_should_be_named_for_the_building_block_and_building_block_type()
+      public void simulations_should_use_modules_from_the_project_to_build_configurations()
       {
-         _context.CurrentProject.Modules.Each(x => x.Name.Contains(x.BuildingBlocks[0].Name).ShouldBeTrue());
-         _context.CurrentProject.Modules.Each(x => x.Name.Contains(_objectTypeResolver.TypeFor(x.BuildingBlocks[0]).Replace("Building Block", string.Empty)).ShouldBeTrue());
+         _context.CurrentProject.Simulations.Each(matchSimulationAndProjectModules);
+      }
+
```
Comment (rwmcintosh) [inline tests/MoBi.Tests/IntegrationTests/ProjectTaskIntegrationTests.cs:64]: remove extra space

**Review by rwmcintosh — CHANGES_REQUESTED** [review]
Good stuff. There's two methods that I think should be renamed and some extra whitespace.

File: tests/MoBi.Tests/IntegrationTests/ProjectTaskIntegrationTests.cs (hunk 9 above)
Comment (benjaperez1983) [reply tests/MoBi.Tests/IntegrationTests/ProjectTaskIntegrationTests.cs:64]: image
I think there is no extra space here, just normal spacing between methods.
Could you confirm?

Comment (rwmcintosh) [reply tests/MoBi.Tests/IntegrationTests/ProjectTaskIntegrationTests.cs:64]: oh yeah. just in the PR, not in the file.

**Review by rwmcintosh — APPROVED** [review]
Just looks like extra spaces in the test files.

There's a couple failing tests in the CI too

**Review by rwmcintosh — APPROVED** [review]

File: src/MoBi.Presentation/Tasks/Interaction/InitialConditionsTask.cs [hunk 10]
```diff
@@ -1,14 +1,17 @@
...
 using MoBi.Core.Commands;
 using MoBi.Core.Domain.Builder;
 using MoBi.Core.Domain.Extensions;
 using MoBi.Core.Domain.Services;
+using MoBi.Core.Events;
 using MoBi.Core.Services;
 using MoBi.Presentation.DTO;
 using MoBi.Presentation.Mappers;
 using MoBi.Presentation.Tasks.Edit;
+using NPOI.POIFS.Properties;
```
Comment (msevestre) [inline src/MoBi.Presentation/Tasks/Interaction/InitialConditionsTask.cs:14]: why was this added?

**Review by msevestre — APPROVED** [review]
